File: STABLE/3.0.0/java/src/org/mariuszgromada/math/mxparser/mathcollection/PrimesCache.java
/*
 * @(#)PrimesCache.java        3.0.0    2016-05-07
 *
 * You may use this software under the condition of "Simplified BSD License"
 *
 * Copyright 2010-2016 MARIUSZ GROMADA. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are
 * permitted provided that the following conditions are met:
 *
 *    1. Redistributions of source code must retain the above copyright notice, this list of
 *       conditions and the following disclaimer.
 *
 *    2. Redistributions in binary form must reproduce the above copyright notice, this list
 *       of conditions and the following disclaimer in the documentation and/or other materials
 *       provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY <MARIUSZ GROMADA> ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are those of the
 * authors and should not be interpreted as representing official policies, either expressed
 * or implied, of MARIUSZ GROMADA.
 *
 * If you have any questions/bugs feel free to contact:
 *
 *     Mariusz Gromada
 *     mariuszgromada.org@gmail.com
 *     http://mathparser.org
 *     http://mathspace.pl
 *     http://janetsudoku.mariuszgromada.org
 *     http://github.com/mariuszgromada/MathParser.org-mXparser
 *     http://mariuszgromada.github.io/MathParser.org-mXparser
 *     http://mxparser.sourceforge.net
 *     http://bitbucket.org/mariuszgromada/mxparser
 *     http://mxparser.codeplex.com
 *     http://github.com/mariuszgromada/Janet-Sudoku
 *     http://janetsudoku.codeplex.com
 *     http://sourceforge.net/projects/janetsudoku
 *     http://bitbucket.org/mariuszgromada/janet-sudoku
 *     http://github.com/mariuszgromada/MathParser.org-mXparser
 *
 *                              Asked if he believes in one God, a mathematician answered:
 *                              "Yes, up to isomorphism."
 */
package org.mariuszgromada.math.mxparser.mathcollection;
/**
 * Class for generating prime numbers cache using
 * Eratosthenes Sieve.
 *
 * @author         <b>Mariusz Gromada</b><br>
 *                 <a href="mailto:mariuszgromada.org@gmail.com">mariuszgromada.org@gmail.com</a><br>
 *                 <a href="http://mathspace.pl" target="_blank">MathSpace.pl</a><br>
 *                 <a href="http://mathparser.org" target="_blank">MathParser.org - mXparser project page</a><br>
 *                 <a href="http://github.com/mariuszgromada/MathParser.org-mXparser" target="_blank">mXparser on GitHub</a><br>
 *                 <a href="http://mxparser.sourceforge.net" target="_blank">mXparser on SourceForge</a><br>
 *                 <a href="http://bitbucket.org/mariuszgromada/mxparser" target="_blank">mXparser on Bitbucket</a><br>
 *                 <a href="http://mxparser.codeplex.com" target="_blank">mXparser on CodePlex</a><br>
 *                 <a href="http://janetsudoku.mariuszgromada.org" target="_blank">Janet Sudoku - project web page</a><br>
 *                 <a href="http://github.com/mariuszgromada/Janet-Sudoku" target="_blank">Janet Sudoku on GitHub</a><br>
 *                 <a href="http://janetsudoku.codeplex.com" target="_blank">Janet Sudoku on CodePlex</a><br>
 *                 <a href="http://sourceforge.net/projects/janetsudoku" target="_blank">Janet Sudoku on SourceForge</a><br>
 *                 <a href="http://bitbucket.org/mariuszgromada/janet-sudoku" target="_blank">Janet Sudoku on BitBucket</a><br>
 *
 * @version        3.0.0
 */
public class PrimesCache {
	/**
	 * Default range of integer to store in cache
	 */
	public static final int DEFAULT_MAX_NUM_IN_CACHE = 10000000;
	/**
	 * Empty cache status
	 */
	public static final boolean CACHE_EMPTY = false;
	/**
	 * Cache ready to use
	 */
	public static final boolean CACHING_FINISHED = true;
	/**
	 * Indicator if given number is a prime
	 */
	public static final int IS_PRIME = 1;
	/**
	 * Indicator if given number is not a prime
	 */
	public static final int IS_NOT_PRIME = 0;
	/**
	 * Indicator that the value is not stored
	 * in cache
	 */
	public static final int NOT_IN_CACHE = -1;
	/**
	 * Primes between 0 ... and ... maximumNumberInCache
	 * will be cached
	 */
	int maxNumInCache;
	/**
	 * Number of cached prime numbers
	 */
	int numberOfPrimes;
	/**
	 * Time in seconds showing
	 * how long did it take to finalize prime numbers
	 * caching.
	 */
	double computingTime;
	/**
	 * Caching process status
	 */
	boolean cacheStatus;
	/*
	 * Integers table to store number and indicate
	 * whether they are prime or not
	 */
	boolean[] isPrime;
	/**
	 * Eratosthenes Sieve implementation
	 */
	private void EratosthenesSieve() {
		isPrime = new boolean[maxNumInCache+1];
		numberOfPrimes = 0;
		long startTime = System.currentTimeMillis();
		/*
		 * Initially assume all integers are primes
		 */
		isPrime[0] = false;
		isPrime[1] = false;
		for (int i = 2; i <= maxNumInCache; i++)
			isPrime[i] = true;
		/*
		 * Sieve of Eratosthenes - marking non-primes
		 */
        for (int i = 2; i*i <= maxNumInCache; i++)
            if (isPrime[i] == true)
                for (int j = i; i*j <= maxNumInCache; j++)
                    isPrime[i*j] = false;
		long endTime = System.currentTimeMillis();
		computingTime = (endTime - startTime)/1000.0;
	}
	/**
	 * Counting found primes
	 */
	private void countPrimes() {
		for (int i = 0; i <= maxNumInCache; i++)
			if (isPrime[i] == true) numberOfPrimes++;
	}
	/**
	 * Default constructor - setting prime cache for a default range if integers
	 */
	public PrimesCache() {
		cacheStatus = CACHE_EMPTY;
		maxNumInCache = DEFAULT_MAX_NUM_IN_CACHE;
		EratosthenesSieve();
		countPrimes();
		cacheStatus = CACHING_FINISHED;
	}
	/**
	 * Constructor - setting prime cache for a given range if integers
	 * @param maxNumInCache Range of integers to be stored in prime cache
	 */
	public PrimesCache(int maxNumInCache) {
		if (maxNumInCache > 2)
			this.maxNumInCache = maxNumInCache;
		else
			this.maxNumInCache = DEFAULT_MAX_NUM_IN_CACHE;
		cacheStatus = CACHE_EMPTY;
		maxNumInCache = DEFAULT_MAX_NUM_IN_CACHE;
		EratosthenesSieve();
		countPrimes();
		cacheStatus = CACHING_FINISHED;
	}
	/**
	 * Returns computing time of Eratosthenes Sieve
	 * @return Computing time in seconds
	 */
	public double getComputingTime() {
		return computingTime;
	}
	/**
	 * Returns cache status
	 * @return PrimesCache.CACHE_EMPTY or PrimesCache.CACHING_FINISHED;
	 */
	public boolean getCacheStatus() {
		return cacheStatus;
	}
	/**
	 * Returns number of found primes.
	 * @return Number of found primes.
	 */
	public int getNumberOfPrimes() {
		return numberOfPrimes;
	}
	/**
	 * Returns cache range.
	 * @return Maximum integera number in cache/
	 */
	public int getMaxNumInCache() {
		return maxNumInCache;
	}
	/**
	 * Check whether given number is prime
	 * @param n Given integer number.
	 * @return PrimesCache.IS_PRIME or PrimesCache.IS_NOT_PRIME or PrimesCache.NOT_IN_CACHE
	 */
	public int primeTest(int n) {
		if ( (n <= maxNumInCache) && (cacheStatus = CACHING_FINISHED) )
			if ( isPrime[n] == true)
				return IS_PRIME;
			else
				return IS_NOT_PRIME;
		else
			return NOT_IN_CACHE;
	}
	/**
	 * Gets underlying primes cache boolean table
	 * @return Underlying primes cache boolean table
	 */
	boolean[] getPrimes() {
		return isPrime;
	}
}

File: CURRENT/typescript/org/mariuszgromada/math/mxparser/mathcollection/PrimesCache.ts
import { mXparserConstants } from '../mXparserConstants';
import { java } from 'j4ts/j4ts';
import { javaemul } from 'j4ts/j4ts';

/**
 * Constructor - setting prime cache for a given range if integers
 * @param {number} maxNumInCache Range of integers to be stored in prime cache
 * @class
 */
export class PrimesCache {
    /**
     * Default range of integer to store in cache
     */
    public static DEFAULT_MAX_NUM_IN_CACHE: number = 10000000;

    /**
     * Empty cache status
     */
    public static CACHE_EMPTY: boolean = false;

    /**
     * Cache ready to use
     */
    public static CACHING_FINISHED: boolean = true;

    /**
     * Indicator if given number is a prime
     */
    public static IS_PRIME: number = 1;

    /**
     * Indicator if given number is not a prime
     */
    public static IS_NOT_PRIME: number = 0;

    /**
     * Indicator that the value is not stored
     * in cache
     */
    public static NOT_IN_CACHE: number = -1;

    /**
     * Primes between 0 ... and ... maximumNumberInCache
     * will be cached
     */
    maxNumInCache: number;

    /**
     * Number of cached prime numbers
     */
    numberOfPrimes: number;

    /**
     * Time in seconds showing
     * how long did it take to finalize prime numbers
     * caching.
     */
    computingTime: number;

    /**
     * Caching process status
     */
    cacheStatus: boolean;

    /**
     * Integers table to store number and indicate
     * whether they are prime or not
     */
    isPrime: boolean[];

    /**
     * Internal flag marking that primes cache initialization was successful;
     */
    initSuccessful: boolean;

    /**
     * Eratosthenes Sieve implementation
     * @private
     */
    /*private*/ EratosthenesSieve() {
        const startTime: number = java.lang.System.currentTimeMillis();
        try {
            const size: number = this.maxNumInCache + 1;
            if (size <= 0){
                this.numberOfPrimes = 0;
                this.maxNumInCache = 0;
                this.initSuccessful = false;
                const endTime: number = java.lang.System.currentTimeMillis();
                this.computingTime = (endTime - startTime) / 1000.0;
                return;
            }
            this.isPrime = (s => { let a=[]; while(s-->0) a.push(false); return a; })(size);
            this.numberOfPrimes = 0;
            this.isPrime[0] = false;
            this.isPrime[1] = false;
            for(let i: number = 2; i <= this.maxNumInCache; i++) {{
                this.isPrime[i] = true;
                if (mXparserConstants.isCurrentCalculationCancelled())return;
            };}
            for(let i: number = 2; i * i <= this.maxNumInCache; i++) {{
                if (mXparserConstants.isCurrentCalculationCancelled())return;
                if (this.isPrime[i] === true)for(let j: number = i; i * j <= this.maxNumInCache; j++) {{
                    this.isPrime[i * j] = false;
                    if (mXparserConstants.isCurrentCalculationCancelled())return;
                };}
            };}
            this.initSuccessful = true;
        } catch(e) {
            this.initSuccessful = false;
        } finally {
            const endTime: number = java.lang.System.currentTimeMillis();
            this.computingTime = (endTime - startTime) / 1000.0;
        }
    }

    /**
     * Counting found primes
     * @private
     */
    /*private*/ countPrimes() {
        for(let i: number = 0; i <= this.maxNumInCache; i++) {{
            if (this.isPrime[i] === true)this.numberOfPrimes++;
            if (mXparserConstants.isCurrentCalculationCancelled())return;
        };}
    }

    public constructor(maxNumInCache?: any) {
        if (((typeof maxNumInCache === 'number') || maxNumInCache === null)) {
            let __args = arguments;
            if (this.maxNumInCache === undefined) { this.maxNumInCache = 0; } 
            if (this.numberOfPrimes === undefined) { this.numberOfPrimes = 0; } 
            if (this.computingTime === undefined) { this.computingTime = 0; } 
            if (this.cacheStatus === undefined) { this.cacheStatus = false; } 
            if (this.isPrime === undefined) { this.isPrime = null; } 
            if (this.initSuccessful === undefined) { this.initSuccessful = false; } 
            if (maxNumInCache > 2)this.maxNumInCache = Math.min(maxNumInCache, javaemul.internal.IntegerHelper.MAX_VALUE - 1); else this.maxNumInCache = PrimesCache.DEFAULT_MAX_NUM_IN_CACHE;
            this.initSuccessful = false;
            this.cacheStatus = PrimesCache.CACHE_EMPTY;
            maxNumInCache = PrimesCache.DEFAULT_MAX_NUM_IN_CACHE;
            this.EratosthenesSieve();
            if (this.initSuccessful){
                this.countPrimes();
                this.cacheStatus = PrimesCache.CACHING_FINISHED;
            } else {
                maxNumInCache = 0;
                this.numberOfPrimes = 0;
            }
        } else if (maxNumInCache === undefined) {
            let __args = arguments;
            if (this.maxNumInCache === undefined) { this.maxNumInCache = 0; } 
            if (this.numberOfPrimes === undefined) { this.numberOfPrimes = 0; } 
            if (this.computingTime === undefined) { this.computingTime = 0; } 
            if (this.cacheStatus === undefined) { this.cacheStatus = false; } 
            if (this.isPrime === undefined) { this.isPrime = null; } 
            if (this.initSuccessful === undefined) { this.initSuccessful = false; } 
            this.initSuccessful = false;
            this.cacheStatus = PrimesCache.CACHE_EMPTY;
            this.maxNumInCache = PrimesCache.DEFAULT_MAX_NUM_IN_CACHE;
            this.EratosthenesSieve();
            if (this.initSuccessful){
                this.countPrimes();
                this.cacheStatus = PrimesCache.CACHING_FINISHED;
            } else {
                this.maxNumInCache = 0;
                this.numberOfPrimes = 0;
            }
        } else throw new Error('invalid overload');
    }

    /**
     * Returns computing time of Eratosthenes Sieve
     * @return {number} Computing time in seconds
     */
    public getComputingTime(): number {
        return this.computingTime;
    }

    /**
     * Returns cache status
     * @return {boolean} PrimesCache.CACHE_EMPTY or PrimesCache.CACHING_FINISHED;
     */
    public getCacheStatus(): boolean {
        return this.cacheStatus;
    }

    /**
     * Returns number of found primes.
     * @return {number} Number of found primes.
     */
    public getNumberOfPrimes(): number {
        return this.numberOfPrimes;
    }

    /**
     * Returns cache range.
     * @return {number} Maximum integera number in cache/
     */
    public getMaxNumInCache(): number {
        return this.maxNumInCache;
    }

    /**
     * Check whether given number is prime
     * @param {number} n Given integer number.
     * @return {number} PrimesCache.IS_PRIME or PrimesCache.IS_NOT_PRIME or PrimesCache.NOT_IN_CACHE
     */
    public primeTest(n: number): number {
        if (n <= 1)return PrimesCache.IS_NOT_PRIME;
        if ((n <= this.maxNumInCache) && (this.cacheStatus = PrimesCache.CACHING_FINISHED))if (this.isPrime[n] === true)return PrimesCache.IS_PRIME; else return PrimesCache.IS_NOT_PRIME; else return PrimesCache.NOT_IN_CACHE;
    }

    /**
     * Returns true in case when primes cache initialization was successful,
     * otherwise returns false.
     * 
     * @return {boolean} Returns true in case when primes cache initialization was successful,
     * otherwise returns false.
     */
    public isInitSuccessful(): boolean {
        return this.initSuccessful;
    }

    /**
     * Gets underlying primes cache boolean table
     * @return {boolean[]} Underlying primes cache boolean table
     */
    getPrimes(): boolean[] {
        return this.isPrime;
    }
}
PrimesCache["__class"] = "org.mariuszgromada.math.mxparser.mathcollection.PrimesCache";